doc/man/gnunet-namestore-dbtool.1
.\" This file is part of GNUnet.
.\" Copyright (C) 2001-2022 GNUnet e.V.
.\"
.\" Permission is granted to copy, distribute and/or modify this document
.\" under the terms of the GNU Free Documentation License, Version 1.3 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.  A
.\" copy of the license is included in the file
.\" FDL-1.3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/fdl.html}.
.\"
.\" Alternately, this document is also available under the General
.\" Public License, version 3 or later, as published by the Free Software
.\" Foundation.  A copy of the license is included in the file
.\" GPL3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/gpl.html
.\"
.\" SPDX-License-Identifier: GPL3.0-or-later OR FDL1.3-or-later
.\"
.Dd September 30, 2022
.Dt GNUNET-NAMESTORE-DBTOOL 1
.Os
.Sh NAME
.Nm gnunet-namestore-dbtool
.Nd manipulate GNU Name System (GNS) namestore database
.Sh SYNOPSIS
.Nm
.Op Fl i | -initialize
.Op Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
.Op Fl r | -reset
.Op Fl h | -help
.Op Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
.Op Fl p Ar PLUGIN | Fl -plugin= Ns Ar PLUGIN
.Op Fl v | -version
.Sh DESCRIPTION
.Nm
can be used to initialize or reset the namestore database.
.Pp
The options are as follows:
.Bl -tag -width indent
.It Fl i | -initialize
Desired operation is initialization of a database.
.It Fl p Ar PLUGINNAME | Fl -plugin= Ns Ar PLUGINNAME
Use the plugin PLUGINNAME.
.It Fl r | -reset
Desired operation is reset and reinitialization of a database
.It Fl h | -help
Print short help on options.
.It Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
Use LOGLEVEL for logging.
Valid values are DEBUG, INFO, WARNING and ERROR.
.It Fl v | -version
Print GNUnet version number.
.El
.\".Sh EXAMPLES
.\".Sh FILES
.Sh SEE ALSO
.Xr gnunet-gns 1 ,
.Xr gnunet-namestore 1
.sp
The full documentation for GNUnet is maintained as a Texinfo manual.
If the
.Xr info 1
and gnunet programs are properly installed at your site, the command
.Pp
.Dl info gnunet
.Pp
should give you access to the complete handbook,
.Pp
.Dl info gnunet-c-tutorial
.Pp
will give you access to a tutorial for developers.
.sp
Depending on your installation, this information is also available in
.Xr gnunet 7 and
.Xr gnunet-c-tutorial 7 .
.\".Sh HISTORY
.\".Sh AUTHORS
.Sh BUGS
Report bugs by using
.Lk https://bugs.gnunet.org
or by sending electronic mail to
.Aq Mt gnunet-developers@gnu.org .


src/namestore/gnunet-namestore-dbtool.c
/*
     This file is part of GNUnet.
     Copyright (C) 2012, 2013, 2014, 2019, 2022 GNUnet e.V.

     GNUnet is free software: you can redistribute it and/or modify it
     under the terms of the GNU Affero General Public License as published
     by the Free Software Foundation, either version 3 of the License,
     or (at your option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     Affero General Public License for more details.

     You should have received a copy of the GNU Affero General Public License
     along with this program.  If not, see <http://www.gnu.org/licenses/>.

     SPDX-License-Identifier: AGPL3.0-or-later
 */
/**
 * @file gnunet-namestore-dbtool.c
 * @brief command line tool to manipulate the database backends for the namestore
 * @author Martin Schanzenbach
 *
 */
#include "platform.h"
#include <gnunet_util_lib.h>
#include <gnunet_namestore_plugin.h>

/**
 * Name of the plugin argument
 */
static char *pluginname;

/**
 * Reset argument
 */
static int reset;

/**
 * Initialize argument
 */
static int init;

/**
 * Return code
 */
static int ret = 0;

/**
 * Task run on shutdown.  Cleans up everything.
 *
 * @param cls unused
 */
static void
do_shutdown (void *cls)
{
  (void) cls;
  if (NULL != pluginname)
    GNUNET_free (pluginname);
}
/**
 * Main function that will be run.
 *
 * @param cls closure
 * @param args remaining command-line arguments
 * @param cfgfile name of the configuration file used (for saving, can be NULL!)
 * @param cfg configuration
 */
static void
run (void *cls,
     char *const *args,
     const char *cfgfile,
     const struct GNUNET_CONFIGURATION_Handle *cfg)
{
  char *db_lib_name;
  char *emsg;
  struct GNUNET_NAMESTORE_PluginFunctions *plugin;

  (void) cls;
  (void) args;
  (void) cfgfile;
  if (NULL != args[0])
    GNUNET_log (
      GNUNET_ERROR_TYPE_WARNING,
      _ ("Superfluous command line arguments (starting with `%s') ignored\n"),
      args[0]);

  GNUNET_SCHEDULER_add_shutdown (&do_shutdown, (void *) cfg);
  if (NULL == pluginname)
  {
    fprintf (stderr, "No plugin given!\n");
    ret = 1;
    GNUNET_SCHEDULER_shutdown ();
    return;
  }
  GNUNET_asprintf (&db_lib_name, "libgnunet_plugin_namestore_%s", pluginname);
  plugin = GNUNET_PLUGIN_load (db_lib_name, (void *) cfg);
  if (NULL == plugin)
  {
    fprintf (stderr, "Failed to load %s!\n", db_lib_name);
    ret = 1;
    GNUNET_SCHEDULER_shutdown ();
    GNUNET_free (db_lib_name);
    return;
  }
  if (reset)
  {
    if (GNUNET_OK != plugin->reset_database (plugin->cls, &emsg))
    {
      // FIXME do we want to return a reason?
      fprintf (stderr, "Failed to reset database: %s\n",
               emsg);
      ret = 1;
      GNUNET_free (emsg);
      GNUNET_free (db_lib_name);
      GNUNET_SCHEDULER_shutdown ();
      return;
    }
  }
  else if (init)
  {
    if (GNUNET_OK != plugin->initialize_database (plugin->cls, &emsg))
    {
      // FIXME do we want to return a reason?
      fprintf (stderr, "Failed to initialize database: %s\n",
               emsg);
      ret = 1;
      GNUNET_free (emsg);
      GNUNET_free (db_lib_name);
      GNUNET_SCHEDULER_shutdown ();
      return;
    }
  }
  GNUNET_SCHEDULER_shutdown ();
  GNUNET_break (NULL == GNUNET_PLUGIN_unload (db_lib_name, plugin));
  GNUNET_free (db_lib_name);
}


/**
 * The main function for gnunet-namestore-dbtool.
 *
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return 0 ok, 1 on error
 */
int
main (int argc, char *const *argv)
{
  struct GNUNET_GETOPT_CommandLineOption options[] =
  { GNUNET_GETOPT_option_flag ('i', "init", gettext_noop (
                                 "initialize database"), &init),
    GNUNET_GETOPT_option_flag ('r',
                               "reset",
                               gettext_noop (
                                 "reset database (DANGEROUS: All existing data is lost!"),
                               &reset),
    GNUNET_GETOPT_option_string (
      'p',
      "plugin",
      "PLUGIN",
      gettext_noop (
        "the namestore plugin to work with, e.g. 'sqlite'"),
      &pluginname),
    GNUNET_GETOPT_OPTION_END };
  int lret;

  if (GNUNET_OK != GNUNET_STRINGS_get_utf8_args (argc, argv, &argc, &argv))
    return 2;

  GNUNET_log_setup ("gnunet-namestore-dbtool", "WARNING", NULL);
  if (GNUNET_OK !=
      (lret = GNUNET_PROGRAM_run (argc,
                                  argv,
                                  "gnunet-namestore-dbtool",
                                  _ (
                                    "GNUnet namestore database manipulation tool"),
                                  options,
                                  &run,
                                  NULL)))
  {
    GNUNET_free_nz ((void *) argv);
    return lret;
  }
  GNUNET_free_nz ((void *) argv);
  return ret;
}
